### ISSUE TITLE
No way to run tests without scikit-bio directory

### ISSUE THREAD

**Issue by anderspitman** [biocore/scikit-bio#844]
One advantage of using nosetests directly, instead of wrapping with a makefile as we're trying to put in place now, is that you can run ``nosetests skbio``, and as long as skbio is installed for your interpreter it will run the tests, ie it doesn't require an skbio source directory. We're going to need some way for users to run tests after doing ``pip install`` or ``conda install``

**Comment by ElDeveloper** [comment]
I think users can do: import skbio; skbio.test() Which should run the
full test suite. However I'm not sure if the doctests are being executed
when .test() is called.

**Comment by anderspitman** [comment]
@ElDeveloper that's exactly what I was just wondering about. I'll play around with it and see if there are any limitations.

**Comment by anderspitman** [comment]
Ok looks like
```
import skbio
skbio.test()
```
does work (if I break a test it catches it), but it doesn't run the doctests.

**Comment by anderspitman** [comment]
Alright, I think we should be able to accomplish this using some combination of ``numpy.testing.Tester`` (what we're currently doing) and ``numpy.testing.rundocs``. There's a ``doctests`` named parameter to ``Tester.test``, but it doesn't seem to actually run the doctests. I'll dig into this and see what I can learn. Assuming we get this working, here's where we'd be:

1) Developers would be instructed to run ``make test``, which runs nosetests as well as pep8, etc.
2) Users would be instructed to import skbio and run ``skbio.test()``, which we'll set up to run all the tests and doctests.

Does that seem good?

**Comment by anderspitman** [comment]
Ah of course `numpy.testing.Tester.test`` isn't running the doctests. That's because of the same old problem of ``nose`` ignoring files.

**Comment by jairideout** [comment]
@anderspitman that sounds good, thanks for looking into this! If we can get all tests and doctests to run programmatically, the instructions (for users) can be something like:

```
$ python -c "import skbio; skbio.test()"
```

**Comment by anderspitman** [comment]
Check it out. ``numpy.testing.Tester.test`` includes an ``extra_argv`` parameter, which allows us to pass args through to nose:

```python
nose_args = ['-I', 'DO_NOT_IGNORE_ANYTHING']
skbio.test(doctests=True, extra_argv=nose_args)
```

That seems to run our tests properly, including doctests. I propose adding a thin wrapper procedure to ``skbio.__init__`` such as the following closure:

```python
tester = Tester()
nose_args = ['-I', 'DO_NOT_IGNORE_ANYTHING']
def run_tests():
    tester.test(doctests=True, extra_argv=nose_args)
test = run_tests
```

So at that point the docs would instruct the user to run exactly what @jairideout suggested.

If this sounds good I can roll it into #808 since the docs are broken for that until this gets resolved.

In the long run it might also be useful to unify the testing functionality behind something like this, assuming we can interact with pep8 and flake through python. That would remove the need for a Makefile.

**Comment by ElDeveloper** [comment]
This is awesome :+1: to this! It would also be a good idea to add a note
to the Contributing.md file describing that doctests are also executed
when calling `skbio.test()`.

**Comment by jairideout** [comment]
Thanks @anderspitman, this looks great! The only potential issue is that each subpackage's `__init__.py` has the same numpy testing code in it, so those will need updating too. Maybe the wrapper function could live in `skbio.util` (similar to `get_data_path`, which is also used for testing)?

:+1: for adding these changes to #808.

**Comment by anderspitman** [comment]
It turns out ``numpy.testing.Tester`` isn't equivalent to running ``nosetests`` from the command line. This can be illustrated by running:

```bash
nosetests --with-doctest -I DO_NOT_IGNORE skbio/tree
```
which passes, and comparing it to

```python
import skbio
skbio.tree.test(doctests=True, extra_argv=['-I', 'DO_NOT_IGNORE'])
```

Which fails. I'm still trying to understand exactly what is different. The way it parses some of the test comparisons doesn't seem to work the same.

So technically our use of ``numpy.testing.Tester`` is currently broken, since it's not actually running any doctests, and if it were it would be failing.

I think it would be ideal to have users run ``nosetests skbio`` after pip installing. If we can find a way to ship our nosetests configuration with scikit-bio, that should do it. I'm trying that route [via this stack overflow questions](http://stackoverflow.com/q/28623609/943814).

Failing that I may be able to get ``numpy.testing.Tester`` to work. It's only broken for a few of our doctests. Mostly it works. Worst case scenario would be us needing to wrap ``nose`` ourselves.

**Comment by jairideout** [comment]
Thanks for investigating this @anderspitman! I agree it'd be nice to be able to run the unit tests after a pip-install. Can you post the output of a failing doctest via `numpy.testing.Tester`?

In the meantime (to get #808 merged), we can probably just update the docs to instruct developers to run `make test` when developing, and then update the user-facing docs to have the long `nosetests` command. Not ideal but that'd at least let developers take advantage of `make test` in the meantime. Does that sound like a reasonable plan?

**Comment by anderspitman** [comment]
Here's a good example. It looks like the numpy tester is splitting the strings into tuples or something:

```
======================================================================
FAIL: Doctest: skbio.tree._tree.TreeNode.to_taxonomy
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/anders/anaconda/envs/skbio/lib/python2.7/doctest.py", line 2226, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for skbio.tree._tree.TreeNode.to_taxonomy
  File "skbio/tree/_tree.py", line 1923, in to_taxonomy

----------------------------------------------------------------------
File "skbio/tree/_tree.py", line 1965, in skbio.tree._tree.TreeNode.to_taxonomy
Failed example:
    for name, lineage in lineages:
        print(name, '; '.join(lineage))
Expected:
    1 Bacteria; Firmicutes; Clostridia
    2 Bacteria; Firmicutes; Bacilli
    3 Bacteria; Bacteroidetes; Sphingobacteria
    4 Archaea; Euryarchaeota; Thermoplasmata
    5 Archaea; Euryarchaeota; Thermoplasmata
    6 Archaea; Euryarchaeota; Halobacteria
    7 Archaea; Euryarchaeota; Halobacteria
    8 Bacteria; Bacteroidetes; Sphingobacteria
    9 Bacteria; Bacteroidetes; Cytophagia
Got:
    ('1', 'Bacteria; Firmicutes; Clostridia')
    ('2', 'Bacteria; Firmicutes; Bacilli')
    ('3', 'Bacteria; Bacteroidetes; Sphingobacteria')
    ('4', 'Archaea; Euryarchaeota; Thermoplasmata')
    ('5', 'Archaea; Euryarchaeota; Thermoplasmata')
    ('6', 'Archaea; Euryarchaeota; Halobacteria')
    ('7', 'Archaea; Euryarchaeota; Halobacteria')
    ('8', 'Bacteria; Bacteroidetes; Sphingobacteria')
    ('9', 'Bacteria; Bacteroidetes; Cytophagia')
```

**Comment by anderspitman** [comment]
And I think I have a working wrapper around nose. It's actually very simple. I think numpy had to add a bunch of stuff specific to their environment, but all we really need is to run basic ``nosetests`` with doctests and ignore. I've testing my implementation in isolation. I'm integrating it into skbio now.

**Comment by jairideout** [comment]
Sounds good, thanks!

**Comment by Jorge-C** [comment]
This might have been discussed before, but keep in mind that doctests only run on Py2 (they don't work on Py3 because of representation issues) so we wouldn't like users/developers to think things are broken in the new Python versions :)

**Comment by anderspitman** [comment]
I actually didn't know that! We can either include a parameter or detect py2 vs py3.

**Comment by Jorge-C** [comment]
That'd be great, thanks @anderspitman!

**Comment by anderspitman** [comment]
@jairideout 2 questions:

1) Would you prefer auto detect py2 vs py3 or explicit doctest parameter for skbio.test()?

2) Do you want me to roll that functionality into #808 or open a separate issue?

**Comment by ebolyen** [comment]
I think autodetection would be ideal, as we really do want to run the doctests whenever we are testing in a python2.7 env. I don't know how difficult that is to do however.

**Comment by jairideout** [comment]
Agree with @ebolyen that autodetection would be really nice. I think this makes sense to add to #808 so that Py3 users don't think that scikit-bio is broken if they test their installation. It should be pretty easy to detect the Python version by using `sys.version_info`. @anderspitman does that sound good to you? If it ends up being a large amount of work we can also tackle it after #808 is merged.

**Comment by anderspitman** [comment]
Yup, I was leaning that way. Should be easy to add.

**Comment by ElDeveloper** [comment]
Since we only need to know if it is py3k or not, you could use
`from future.utils import PY3`, where `PY3` is a boolean value.

**Comment by anderspitman** [comment]
:+1: for ``future.utils.PY3``

**Comment by jairideout** [comment]
Fixed in #808.
